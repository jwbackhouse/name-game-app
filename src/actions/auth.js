import { firebase, googleAuthProvider } from '../firebase/firebase';

// NOTE this is called in app.js rather than startLogin so that it runs when app first loads, not just when user explictly logs in/out
export const login = (uid) => ({
  type:'LOGIN',
  uid
});

export const startLogin = () => {
  return () => {    // Returns a function for async actions (can be called with dispatch) - allowed via thunk
    return firebase.auth().signInWithPopup(googleAuthProvider)
    // .then((result) => {
    //   const uid = result.user.uid;
    //   dispatch(login(uid));
    // });
  };
};

export const logout = () => ({
    type:'LOGOUT',
});

export const startLogout = () => {
  return () => {
    return firebase.auth().signOut();
  };
};