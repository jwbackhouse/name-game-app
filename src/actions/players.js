// *** ACTION CREATORS ***

import database from '../firebase/firebase';


// Fetch players from Firebase
export const getPlayers = () => {
  return (dispatch, getState) => {
    dispatch(getPlayersBegin());
    return database.ref(`players`).once('value')
      .then(snapshot => {
        const playerArr = [];
        snapshot.forEach(childSnapshot => {
          const output = childSnapshot.val();
          playerArr.push({
            uid: childSnapshot.key,
            ...output
          })
        });
        dispatch(getPlayersSuccess(playerArr));
      })
      .catch(err => dispatch(getPlayersFailure(err)));
  };
};

export const getPlayersBegin = () => ({
  type: 'GET_PLAYERS_BEGIN'
});

export const getPlayersSuccess = playerArr => ({
    type: 'GET_PLAYERS_SUCCESS',
    payload: playerArr
});

export const getPlayersFailure = error => ({
  type: 'GET_PLAYERS_FAILURE',
  payload: {error}
});


// Remove all players
export const removeAllPlayers = () => ({
  type: 'REMOVE_ALL_PLAYERS'
});


// Add player
export const addPlayer = (player) => ({
  type:'ADD_PLAYER',
  player
});


// Mark player as ready to play
export const markPlayerReady = (uid) => {
  return (dispatch, getState) => {
    return database.ref(`players/${uid}/isReady`).set(true);
  }
};

// export const updatePlayers = (players) => ({
//   type: 'UPDATE_PLAYERS',
//   payload: players
// });

