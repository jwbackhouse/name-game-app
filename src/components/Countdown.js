import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import database from '../firebase/firebase';

export class Countdown extends React.Component {
  state = {
    timerOn: false,
    timerTime: 0,
    timerStart: 0,
    timerLength: this.props.game.timerLength * 1000,
    message: '',
  }
  
  componentDidMount = () => {
    // Check for start time on Firebase
    database.ref('game/startTime').on('value', snapshot => {
      const startTime = snapshot.val();
      if (startTime) {
        this.setState({ timerStart: startTime, message: '' });
        this.startTimer();
      } else {
        this.setState({ message: 'Hang on a second...' });
      }
    });
  }
  
  componentWillUnmount = () => {
    database.ref('game/startTime').off();
    this.stopTimer();
  }
  
  startTimer = () => {
    this.setState({
      timerOn: true,
    });
    
    this.timer = setInterval(() => {
      const timeLeft = this.state.timerLength - this.state.timerTime;
      if (timeLeft >= 0)  {
        this.setState({ timerTime: Date.now() - this.state.timerStart });
      } else {
        clearInterval(this.timer);
        this.setState({ timerOn: false });
        this.props.onFinished();
      }
    }, 50);
  }

  stopTimer = () => {
    clearInterval(this.timer);
    this.setState({ timerOn: false });
  };
  
  // // Unused functionality
  // resetTimer = () => {
  //   this.setState({
  //     timerTime: 0,
  //     timerStart: 0
  //   });
  // };
  
  
  render = () => {
    const { timerLength, timerTime, message } = this.state;
    const timeLeft = Math.max(0, timerLength - timerTime);
    const messageElement = message && message;
    return (
      <div>
        <p>{ messageElement }</p>
        <h3>Time left: {moment(timeLeft).format('m:ss')}</h3>
      </div>
    )
  }
};

const mapStateToProps = (state) => ({
  game: state.game,
});

export default connect(mapStateToProps)(Countdown);
