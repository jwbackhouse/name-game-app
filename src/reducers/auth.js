// *** AUTH REDUCER

const initialState = {
  uid: '',
  username: '',
  error: ''
};

export default (state = initialState, action) => {
  switch(action.type) {
    case 'LOGIN_SUCCESS':
      return {
        uid: action.uid,
        username: action.username,
        error: ''
      };
    case 'LOGIN_ERROR':
      return {
        uid: '',
        error: action.error
      };
    case 'LOGOUT_SUCCESS':
      return initialState;
    case 'LOGOUT_FAILURE':
      return {
        ...state,
        error: action.error
      };
    default:
      return state;
  }
};
