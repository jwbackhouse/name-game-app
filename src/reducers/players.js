const initialState = {
  players: [],
  isLoading: false,
  error: false
};

export default (state = initialState, action) => {
  switch(action.type) {
    case 'GET_PLAYERS_BEGIN':
      return {
        ...state,
        isLoading: true,
        error: null
      }
    case 'GET_PLAYERS_SUCCESS':
      return {
        ...state,
        isLoading: false,
        players: action.payload
      }
    case 'GET_PLAYERS_FAILURE':
      return {
        ...state,
        isLoading: false,
        error: action.payload.error
      }
    case 'ADD_PLAYER':
      return {
        ...state,
        players: [...state.players, action.player]
      };
    case 'REMOVE_ALL_PLAYERS':
      return state.players = [];
    case 'UPDATE_PLAYERS':
        return {
          ...state,
          players: action.payload
        }
    default:
      return state;
  }
};
