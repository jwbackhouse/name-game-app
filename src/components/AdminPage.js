import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { Link, useHistory } from 'react-router-dom';
import database from '../firebase/firebase';
import withLiveData from '../helpers/withLiveData';
import { resetGame, initialiseGame } from '../actions/game';
import { removeAllNames } from '../actions/names';


const AdminPage = ({ initialiseGame, removeAllNames, resetGame, game, history }) => {
  const [submitMsg, setSubmitMsg] = useState('');
  
  // Number of passes allowed per player
  const [numPasses, setNumPasses] = useState(game.numPasses || 2);
  const onPassesChange = (e) => setNumPasses(+e.target.value);
  useEffect(() => {
    if (game.numPasses !== undefined) setNumPasses(game.numPasses)  // needed otherwise input becomes uncontrolled
  }, [game.numPasses]);
  
  // Length of timer for guesses
  const [timerLength, setTimerLength] = useState(game.timerLength || 60000);
  const onTimerChange = (e) => {
    const milliseconds = e.target.value * 1000;
    setTimerLength(milliseconds);
  };
  
  // Very basic password protection for page access
  const [access, setAccess] = useState(undefined);
  if (access === undefined ) {
    const password = prompt('Please enter your password to view this page');
    const pass1 = 'namegame';
    password === pass1 ? setAccess(true): setAccess(false);
  }
  
  const onReset = () => {
    const confirm = window.confirm('Are you sure you want to reset the game? This can\'t be undone...');
    if (confirm) {
      database.ref().remove()
        .then(() => {
          initialiseGame();
          setSubmitMsg('Data successfully reset.');
        });
      removeAllNames();
      resetGame();
    }
  }
  
  const onSubmit = (e) => {
    e.preventDefault();
    
    database.ref('game').update({
      numPasses,
      timerLength,
    });
    setSubmitMsg('All saved for you.');
  };
  
  const pageContent = (
    <div>
      <h1>Game settings</h1>
      <form onSubmit={ onSubmit }>
        <label>Number of passes allowed
          <input
            name='numPasses'
            id='numPasses'
            value={ numPasses }
            onChange={ onPassesChange }
            type='number'
          />
        </label>
        <label>Time allowed for guessing (in seconds)
          <input
            name='timerLength'
            id='timerLength'
            value={ timerLength / 1000 }
            onChange={ onTimerChange }
            type='number'
          />
        </label>
        <button type='submit'>
          Save
        </button>
      </form>
      { submitMsg && <p>{ submitMsg }</p> }
      <br />
      <button onClick={ onReset }>
        Reset the game
      </button>
      <br />
      <br />
      <Link
        to='/'
        className='button'
      >
        Go home
      </Link>
    </div>
  );
  
  return access ? pageContent : null;
};

const mapStateToProps = (state) => ({
  game: state.game,
});

const mapDispatchToProps = (dispatch) => ({
  removeAllNames: () => dispatch(removeAllNames()),
  resetGame: () => dispatch(resetGame()),
  initialiseGame: () => dispatch(initialiseGame())
});


const connectedWithLiveData = compose(
  connect(mapStateToProps, mapDispatchToProps),
  withLiveData,
);

export default connectedWithLiveData(AdminPage);
