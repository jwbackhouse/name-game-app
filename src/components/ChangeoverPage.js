import React from 'react';
import { connect } from 'react-redux';
import StartMessage from './StartMessage';
import { fetchData, endFetchData, startTurn } from '../actions/game';
import { getNames } from '../actions/names';
import selectPlayer from '../selectors/selectPlayer';



export class ChangeoverPage extends React.Component {
  state = {
    error: '',
    nextPlayer: undefined
  }
  
  componentDidMount = () => {
    this.props.fetchData();
    this.props.getNames();
  }
  
  componentDidUpdate = (prevProps) => {
    if (!prevProps.game.startTurn) {
      this.props.game.startTurn && this.props.history.push('/guess');
    }
    if (this.props.game.endGame) {
      this.props.history.push('/end')
    }
  }
  
  componentWillUnmount = () => {
    this.props.endFetchData();
  }

  onClick = () => {
    // Update Firebase with startGame & startTime
    this.props.startTurn();
    this.props.history.push('/play')
  }
  
  render = () => {
    // Check if this user is playing next
    const thisUserPlaying = this.props.user.uid === this.props.game.playingNow.uid;

    return (
      <div className='content-container'>
        <h1>The scores on the doors</h1>
        <p>Team A: {this.props.game.teamAScore}</p>
        <p>Team B: {this.props.game.teamBScore}</p>
        <StartMessage
          thisUserPlaying={ thisUserPlaying }
          playingNow={ this.props.game.playingNow }
          errorMsg={ '' }
          onClick={ this.onClick }
        />
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  players: state.players,
  game: state.game,
  user: state.user
});

const mapDispatchToProps = (dispatch) => ({
  startTurn: () => dispatch(startTurn()),
  fetchData: () => dispatch(fetchData()),
  getNames: () => dispatch(getNames()),
  endFetchData: () => dispatch(endFetchData())
});


export default connect(mapStateToProps, mapDispatchToProps)(ChangeoverPage);

