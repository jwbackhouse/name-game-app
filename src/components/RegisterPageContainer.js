import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import RegisterPage from './RegisterPage';
import withLiveData from '../helpers/withLiveData';
import { updateDisplayName, startAddUserDetails } from '../actions/auth';


export const RegisterPageContainer = (props) => {
  const [username, setUsername] = useState(props.auth.username || '');
  const [team, setTeam] = useState('A');
  const [error, setError] = useState('');
  const { players } = props;

  // useEffect(() => setUsername(props.auth.username), [props.auth.username]);
  
  const onNameChange = (e) => setUsername(e.target.value);
  const onTeamChange = (e) => setTeam(e.target.value);
  const onSubmit = (e) => {
    const {
      startAddUserDetails,
      auth,
      updateDisplayName,
      history
    } = props;
    e.preventDefault();

    // Check a name has been entered
    if (!username) {
      setError('^^Please enter your name');
    } else {
      const user = {
        username,
        team
      };
      // Add game-specific details to state.user and push to Firebase>Players
      startAddUserDetails(user);
      // Update displayName in firebase.authUser if it doesn't match
      auth.username !== username && updateDisplayName(username);
      history.push('/setup');
    }
  };

  return (
    <RegisterPage
      error={ error }
      username={ username }
      team={ team }
      players={ players }
      onSubmit={ onSubmit }
      onNameChange={ onNameChange }
      onTeamChange={ onTeamChange }
    />
  );
}

const mapDispatchToProps = (dispatch) => ({
  startAddUserDetails: (user) => dispatch(startAddUserDetails(user)),
  updateDisplayName: (displayName) => dispatch(updateDisplayName(displayName))
});

const mapStateToProps = (state) => ({
  auth: state.auth
})

const connectedWithLiveData = compose(
  connect(mapStateToProps, mapDispatchToProps),
  withLiveData
);

export default connectedWithLiveData(RegisterPageContainer);

// NB alternative syntax for the above 5 lines:
// export default withLiveData(connect(mapStateToProps, mapDispatchToProps)(RegisterPage));
