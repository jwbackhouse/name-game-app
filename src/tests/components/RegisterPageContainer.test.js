import React from 'react';
import { shallow, mount } from 'enzyme';
import auth, { authWithError, blankAuth } from '../fixtures/testAuth';
import players from '../fixtures/testPlayers';
import { RegisterPageContainer} from '../../components/RegisterPageContainer';


describe('<RegisterPageContainer mounted />', () => {
  let wrapper;
  let RegisterPage;
  const updateDisplayName = jest.fn();
  const startAddUserDetails = jest.fn();
  const history = {push: jest.fn()};

  beforeEach(() => {
    wrapper = mount(
      <RegisterPageContainer
        auth={ auth }
        players={ players }
        startAddUserDetails={ startAddUserDetails }
        updateDisplayName={ updateDisplayName }
        history = { history}
      />
    );
    RegisterPage = wrapper.find('RegisterPage');
  });
  
  afterEach(() => {
    jest.clearAllMocks();
  });
  
  
  test('Should render RegisterPageContainer correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  
  test('Should pass initial state values to props', () => {
    expect(RegisterPage.props().username).toEqual(auth.username);
    expect(RegisterPage.props().team).toEqual('A');
  });
    
  test('Should handle username input correctly.', () => {
    const value = 'Brian';
    wrapper.find('input').simulate('change', {
      target: {value}
    });
    expect(wrapper.find('input').prop('value')).toBe(value);
  });
    
  test('Should handle team input correctly.', () => {
    const value = 'B';
    wrapper.find('select').simulate('change', {
      target: {value}
    });
    expect(wrapper.find('select').prop('value')).toBe(value);
  });
  
  test('Should show error message if no username entered', () => {
    const value = '';
    expect(wrapper.find('.error').exists()).toEqual(false);
    wrapper.find('input').simulate('change', {
      target: {value}
    });
    wrapper.find('button').simulate('submit');
    // NB alternative way to simulate form submissions is as follows:
    // wrapper.find('form').simulate('submit', {
    //   preventDefault: () => {}
    // });
    expect(wrapper.find('.error').exists()).toEqual(true);
    expect(startAddUserDetails).not.toBeCalled();
    expect(history.push).not.toBeCalled();
    expect(updateDisplayName).not.toBeCalled();
  });
  
  test('Should call correct functions when form submitted', () => {
    const value = 'Malcolm Tucker';
    wrapper.find('input').simulate('change', {
      target: {value}
    });
    wrapper.find('button').simulate('submit');
    expect(wrapper.find('.error').exists()).toEqual(false);
    expect(startAddUserDetails).toBeCalledWith({ username: value, team: 'A' });
    expect(history.push).toBeCalledWith('/setup');
    expect(updateDisplayName).toBeCalledWith(value);
  });
});
