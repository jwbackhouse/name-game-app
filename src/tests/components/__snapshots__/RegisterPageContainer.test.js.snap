// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterPageContainer mounted /> Should render RegisterPageContainer correctly 1`] = `
<RegisterPageContainer
  auth={
    Object {
      "error": "",
      "passwordResetError": "",
      "playersUid": 1234,
      "uid": "abc1",
      "username": "Joe",
    }
  }
  history={
    Object {
      "push": [MockFunction],
    }
  }
  players={
    Object {
      "error": false,
      "isLoading": false,
      "players": Array [
        Object {
          "team": "A",
          "uid": 1234,
          "username": "Bob",
        },
        Object {
          "team": "B",
          "uid": 5678,
          "username": "Bridget",
        },
        Object {
          "team": "A",
          "uid": 9012,
          "username": "Amy",
        },
      ],
    }
  }
  startAddUserDetails={[MockFunction]}
  updateDisplayName={[MockFunction]}
>
  <RegisterPage
    error=""
    onNameChange={[Function]}
    onSubmit={[Function]}
    onTeamChange={[Function]}
    players={
      Object {
        "error": false,
        "isLoading": false,
        "players": Array [
          Object {
            "team": "A",
            "uid": 1234,
            "username": "Bob",
          },
          Object {
            "team": "B",
            "uid": 5678,
            "username": "Bridget",
          },
          Object {
            "team": "A",
            "uid": 9012,
            "username": "Amy",
          },
        ],
      }
    }
    team="A"
    username="Joe"
  >
    <div
      className="content-container"
    >
      <h1>
        Sign up here
      </h1>
      <form
        onSubmit={[Function]}
      >
        <input
          autoFocus={true}
          className="text-input"
          name="username"
          onChange={[Function]}
          placeholder="Name"
          value="Joe"
        />
        <select
          className="select"
          name="team"
          onChange={[Function]}
          value="A"
        >
          <option
            value="A"
          >
            Team A
          </option>
          <option
            value="B"
          >
            Team B
          </option>
        </select>
        <button
          className="button button--input"
        >
          Go
        </button>
        <div>
          <p
            className="list-header"
          >
            Already on Team A:
          </p>
          <TeamList
            players={
              Object {
                "error": false,
                "isLoading": false,
                "players": Array [
                  Object {
                    "team": "A",
                    "uid": 1234,
                    "username": "Bob",
                  },
                  Object {
                    "team": "B",
                    "uid": 5678,
                    "username": "Bridget",
                  },
                  Object {
                    "team": "A",
                    "uid": 9012,
                    "username": "Amy",
                  },
                ],
              }
            }
            team="A"
          >
            <ul
              className="list"
            >
              <li
                className="list__item"
                key="1234"
              >
                Bob
              </li>
              <li
                className="list__item"
                key="9012"
              >
                Amy
              </li>
            </ul>
          </TeamList>
          <p
            className="list-header"
          >
            Already on Team B:
          </p>
          <TeamList
            players={
              Object {
                "error": false,
                "isLoading": false,
                "players": Array [
                  Object {
                    "team": "A",
                    "uid": 1234,
                    "username": "Bob",
                  },
                  Object {
                    "team": "B",
                    "uid": 5678,
                    "username": "Bridget",
                  },
                  Object {
                    "team": "A",
                    "uid": 9012,
                    "username": "Amy",
                  },
                ],
              }
            }
            team="B"
          >
            <ul
              className="list"
            >
              <li
                className="list__item"
                key="5678"
              >
                Bridget
              </li>
            </ul>
          </TeamList>
        </div>
      </form>
    </div>
  </RegisterPage>
</RegisterPageContainer>
`;
